nixos/modules/virtualisation/qemu-vm.nix

```
@@ -854,7 +854,7 @@ in
         "-device virtio-gpu-pci" "-device usb-ehci,id=usb0" "-device usb-kbd" "-device usb-tablet"
       ])
       (mkIf (!cfg.useBootLoader) [
-        "-kernel ${config.system.build.toplevel}/kernel"
+        "-kernel \${QEMU_KERNEL_${config.system.name}:-${config.system.build.toplevel}/kernel}"
```

- Let's make clear that this isn't qemu behavior.
 - Make sure we don't break on `-` or other funny characters.

```suggestion
        "-kernel \${NIXPKGS_QEMU_KERNEL_${config.system.name}:-${config.system.build.toplevel}/kernel}"
```

Maybe add something like this?

```nix
lib.strings.sanitizeEnvironmentVariableName = lib.strings.stringAsChars (x: if builtins.match "[a-zA-Z0-9_]" x != null then x else "_") "a-b_c"
```

---

Hi,

Regarding:

> lib.strings.sanitizeEnvironmentVariableName = lib.strings.stringAsChars (x: if builtins.match "[a-zA-Z0-9_]" x != null then x else "_")

`config.system.name` serves as VM name. Therefore it is already sanitised at this stage.

Thanks,
Boris.